--- a/drivers/block/xen-blkfront.c
+++ b/drivers/block/xen-blkfront.c
@@ -2371,6 +2371,7 @@
 	unsigned int physical_sector_size;
 	unsigned int binfo;
 	int err, i;
+	int removable;
 
 	switch (info->connected) {
 	case BLKIF_STATE_CONNECTED:
@@ -2417,6 +2418,12 @@
 		return;
 	}
 
+ 	err = xenbus_gather(XBT_NIL, info->xbdev->otherend,
+			    "removable", "%d", &removable,
+			    NULL);
+	if (!err && removable)
+		binfo |= VDISK_REMOVABLE;
+
 	/*
 	 * physcial-sector-size is a newer field, so old backends may not
 	 * provide this. Assume physical sector size to be the same as
